File: app-serv/internal/schema/log.go
// Package schema holds request/response DTOs and their validation rules.
//
// @file      internal/schema/log.go
// @for       The request-log list, detail, purge, and console-buffer contracts.
// @uses      internal/domain (LogFilter, RequestLog) and the shared usage filter.
// @reason    SPEC-API-001 §7.13 fixes the log filters, the captured-body detail,
//
//	and the console buffer shape; §2.4 requires them typed before the
//	handler, and the detail response has to distinguish "capture is
//	off" from "no bodies were stored" for the panel to be honest.
//
// @author    Dodi Rusmana <rusmanadodi@kentangtech.com>
// @layer     schema
// @stability experimental
// @since     2026-09-18
package schema

import (
	"net/http"
	"time"

	"github.com/rusmanadodi2598/pannelAI/app-serv/internal/domain"
)

// LogFilterQuery is the decoded log filter set (SPEC-API-001 §7.13). It reuses
// the usage filter decoder for the shared fields so one request cannot be
// filtered two different ways depending on which route read it.
type LogFilterQuery struct {
	UsageFilterQuery
}

// DecodeLogFilter reads and validates the log filters.
func DecodeLogFilter(r *http.Request) (LogFilterQuery, error) {
	decoded, err := DecodeUsageFilter(r)
	if err != nil {
		return LogFilterQuery{}, err
	}
	return LogFilterQuery{UsageFilterQuery: decoded}, nil
}

// Filter lowers the validated query into the domain filter with the same
// bounded default window the usage reads use.
func (q LogFilterQuery) Filter(now time.Time) domain.LogFilter {
	return domain.NewLogFilter(domain.LogFilterInput{
		From:       q.From,
		To:         q.To,
		Status:     q.Status,
		EndpointID: q.EndpointID,
		Model:      q.Model,
		Query:      q.Q,
		GatewayKey: q.GatewayKey,
	}, now)
}

// LogRecordResponse is one request-log row without its bodies, which is what a
// list needs and what keeps a page from carrying megabytes of payload.
type LogRecordResponse struct {
	RequestID    string `json:"request_id"`
	TS           string `json:"ts"`
	GatewayKeyID string `json:"gateway_key_id,omitempty"`
	EndpointID   string `json:"endpoint_id,omitempty"`
	ProviderID   string `json:"provider_id,omitempty"`
	Model        string `json:"model,omitempty"`
	Status       string `json:"status"`
	LatencyMS    int64  `json:"latency_ms"`
	Error        string `json:"error,omitempty"`
	HasBodies    bool   `json:"has_bodies"`
}

// LogList wraps a page of log rows with the pagination meta block.
type LogList struct {
	Data []LogRecordResponse `json:"data"`
	Meta Page                `json:"meta"`
}

// LogDetailResponse is the full captured request and response. Bodies are
// omitted when capture is off, and each body carries the truncation marker the
// capture path wrote, so a truncated body says so at the point it was cut
// rather than ending silently (SPEC-UI §6.11).
type LogDetailResponse struct {
	RequestID           string `json:"request_id"`
	TS                  string `json:"ts"`
	GatewayKeyID        string `json:"gateway_key_id,omitempty"`
	EndpointID          string `json:"endpoint_id,omitempty"`
	ProviderID          string `json:"provider_id,omitempty"`
	Model               string `json:"model,omitempty"`
	Status              string `json:"status"`
	LatencyMS           int64  `json:"latency_ms"`
	Error               string `json:"error,omitempty"`
	CaptureEnabled      bool   `json:"capture_enabled"`
	CaptureBodyMaxBytes int    `json:"capture_body_max_bytes"`
	RequestBody         string `json:"request_body,omitempty"`
	ResponseBody        string `json:"response_body,omitempty"`
}

// LogPurgeResponse reports how many rows a purge removed.
type LogPurgeResponse struct {
	Deleted int64 `json:"deleted"`
}

// ConsoleResponse is the body of GET /api/v1/logs/console. Lines are oldest
// first so the panel appends them without reversing, and MaxRecords is the
// bound in force so the panel can say why older lines are gone.
type ConsoleResponse struct {
	Lines      []string `json:"lines"`
	MaxRecords int      `json:"max_records"`
}

// LogRecordResponseFrom maps one stored log row onto the wire shape.
func LogRecordResponseFrom(entry domain.RequestLog) LogRecordResponse {
	return LogRecordResponse{
		RequestID:    entry.RequestID(),
		TS:           Timestamp(entry.TS()),
		GatewayKeyID: entry.GatewayKeyID(),
		EndpointID:   entry.EndpointID(),
		ProviderID:   entry.ProviderID(),
		Model:        entry.Model(),
		Status:       entry.Status(),
		LatencyMS:    entry.LatencyMS(),
		Error:        entry.Error(),
		HasBodies:    entry.RequestBody() != "" || entry.ResponseBody() != "",
	}
}

// LogDetailResponseFrom maps one stored log row and the capture settings that
// governed it onto the wire shape. The bodies are dropped entirely when capture
// is off, so a stale body written before capture was disabled cannot be read as
// if it were current.
func LogDetailResponseFrom(entry domain.RequestLog, captureEnabled bool, maxBytes int) LogDetailResponse {
	resp := LogDetailResponse{
		RequestID:           entry.RequestID(),
		TS:                  Timestamp(entry.TS()),
		GatewayKeyID:        entry.GatewayKeyID(),
		EndpointID:          entry.EndpointID(),
		ProviderID:          entry.ProviderID(),
		Model:               entry.Model(),
		Status:              entry.Status(),
		LatencyMS:           entry.LatencyMS(),
		Error:               entry.Error(),
		CaptureEnabled:      captureEnabled,
		CaptureBodyMaxBytes: maxBytes,
	}
	if captureEnabled {
		resp.RequestBody = entry.RequestBody()
		resp.ResponseBody = entry.ResponseBody()
	}
	return resp
}

// ConsoleResponseFrom maps the stored console lines onto the wire shape.
func ConsoleResponseFrom(lines []string, maxRecords int) ConsoleResponse {
	if lines == nil {
		lines = []string{}
	}
	return ConsoleResponse{Lines: lines, MaxRecords: maxRecords}
}

File: app-ui/src/lib/schemas/log.ts
// Request log schemas, mirroring docs/SPEC-API/001-SPEC-API.md §7.13.
//
// Three wire facts shape this file. The list row carries `has_bodies` rather than the bodies themselves,
// so a page of a hundred requests does not carry a hundred payloads. The detail carries the capture
// setting beside the bodies, because "capture is off" and "capture is on with nothing stored" are
// different answers and the operator needs the difference. And the console read carries `max_records`,
// which is the bound in force, so the screen can say why older lines are gone instead of implying the
// buffer always held that much.
//
// The bodies render verbatim in an escaped, read-only block (SPEC-UI §7.5.1). Nothing here rewrites them:
// a stored payload is evidence, and normalizing it would make the audit view lie about what the gateway
// recorded. `capture_body_max_bytes` is carried so the screen can say why a body stops where it does.

import { z } from 'zod';
import {
	nullableList,
	pageMeta,
	rfc3339Timestamp,
	schemaRequestStatus,
	stringList
} from './primitives';

export const schemaLogRecord = z.object({
	request_id: z.string().min(1),
	ts: rfc3339Timestamp,
	gateway_key_id: z.string().optional(),
	endpoint_id: z.string().optional(),
	provider_id: z.string().optional(),
	model: z.string().optional(),
	status: schemaRequestStatus,
	latency_ms: z.number().int().min(0),
	error: z.string().optional(),
	has_bodies: z.boolean()
});

export type LogRecord = z.infer<typeof schemaLogRecord>;

export const schemaLogList = z.object({
	data: nullableList(schemaLogRecord),
	meta: pageMeta
});

export type LogList = z.infer<typeof schemaLogList>;

// The captured request and response of one request. Identical to the shape the usage detail joins, so the
// two screens cannot disagree about what a stored body looks like.
export const schemaLogDetail = z.object({
	request_id: z.string().min(1),
	ts: rfc3339Timestamp,
	gateway_key_id: z.string().optional(),
	endpoint_id: z.string().optional(),
	provider_id: z.string().optional(),
	model: z.string().optional(),
	status: schemaRequestStatus,
	latency_ms: z.number().int().min(0),
	error: z.string().optional(),
	capture_enabled: z.boolean(),
	capture_body_max_bytes: z.number().int().min(0),
	request_body: z.string().optional(),
	response_body: z.string().optional()
});

export type LogDetail = z.infer<typeof schemaLogDetail>;

// The purge reports how many rows it removed. The number is real, so the confirmation can say what
// happened rather than only that something did.
export const schemaLogPurge = z.object({
	deleted: z.number().int().min(0)
});

export type LogPurge = z.infer<typeof schemaLogPurge>;

// The console ring buffer. `lines` is oldest first, which is the order the API sends and the order the
// screen appends, so nothing reverses it. `max_records` is the ceiling that evicted the older lines.
export const schemaConsoleLog = z.object({
	lines: stringList,
	max_records: z.number().int().min(0)
});

export type ConsoleLog = z.infer<typeof schemaConsoleLog>;
